### ISSUE TITLE
Error when updating Next.js to v12

### ISSUE THREAD

**Issue by anttimaki** [preconstruct/preconstruct#444]
I'm having problems updating a Next.js app to version 12 in a monorepo project. I'm not sure whether the problem lies within Preconstruct, config (Babel, TS, Next) or Next itself (new Rust-based compiler), but this felt like the most obvious place to ask for help.

I've created [a repo with a minimal reproduction](https://github.com/anttimaki/preconstruct-nextjs12-problem) of the issue, see the README for more details. Basically, a setup that works with Next.js v11, fails with v12 when using one of two 3rd party libraries (`react-hook-form`, `js-cookie`) in packages (using the same libraries in the Next.js app causes no problems). The problem only occurs **before** Preconstruct is used to build the packages. Loading a page in Next.js server results in a error message:

> Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

The project uses `@preconstruct/cli@2.1.5` and `@preconstruct/next@3.0.1`.

Any thoughts about what might be causing the issue or how to solve it are much appreciated.

**Comment by anttimaki** [comment]
Update: I managed to get things working by falling back to using Next's Babel-based compiler. The following configuration solved the issue:

```
module.exports = {
  presets: [
    "next/babel",
    [
      "@babel/preset-env",
      {
        modules: "cjs",
      },
    ],
  ],
  plugins: ["@babel/plugin-transform-runtime"],
};
```

Honestly I can live with this, but it would be nice to know if it's possible to use the faster Rust-based compiler with the setup I've described in the original post.

**Comment by namjul** [comment]
I encountered the same issue. 

`@preconstruct/next` setups babel to look for babel config files through setting [upward-optional](https://babeljs.io/docs/en/options#rootmode) 

https://github.com/preconstruct/preconstruct/blob/9fc0cf5669a748be0628de70d006094e134096ec/packages/next/index.js#L10

Looking at the code how nextjs checks for [SWC](https://github.com/swc-project/swc) usage here

https://github.com/vercel/next.js/blob/a803638da2483f0c468522d5da1037657af53e05/packages/next/build/webpack-config.ts#L399-L419

it seems that it will not use babel just by using `@preconstruct/next`.

This has been actually mentioned here https://github.com/preconstruct/preconstruct/pull/420/files#diff-0fa1531fad93d278562b73a735ad5baf31a68b4f7403e72a05af3e3239e24a4e

It works if you add:

```js
// apps/nextjs/babel.config.js
module.exports = require("../babel.config");
```

**Comment by namjul** [comment]
> Maybe we should mention this in [main/packages/next/README.md](https://github.com/preconstruct/preconstruct/blob/main/packages/next/README.md?rgh-link-date=2022-02-03T16%3A00%3A33Z)

Add a PR https://github.com/preconstruct/preconstruct/pull/445

**Comment by namjul** [comment]
> Honestly I can live with this, but it would be nice to know if it's possible to use the faster Rust-based compiler with the setup I've described in the original post.

Making it work with SWC is still an issue.

**Comment by namjul** [comment]
Using the following patch it works.

```diff
diff --git a/packages/components/package.json b/packages/components/package.json
index 390e2ed..fd42f2c 100644
--- a/packages/components/package.json
+++ b/packages/components/package.json
@@ -2,8 +2,8 @@
   "name": "components",
   "version": "1.0.0",
   "description": "Example package",
-  "main": "dist/components.cjs.js",
-  "module": "dist/components.esm.js",
+  "main": "src/index.ts",
+  "module": "src/index.ts",
   "dependencies": {
     "@types/react": "^17.0.38",
     "js-cookie": "^3.0.1",
diff --git a/packages/components/src/CookieReader.tsx b/packages/components/src/CookieReader.tsx
index 412256a..0f5c71e 100644
--- a/packages/components/src/CookieReader.tsx
+++ b/packages/components/src/CookieReader.tsx
@@ -5,7 +5,7 @@ export const CookieReader: React.FC = () => {
   let cookie: string | undefined = "not read";
 
   // The next line works with Next.js v11, but breaks with v12.
-  // cookie = Cookies.get("someCookie");
+  cookie = Cookies.get("someCookie");
 
   return <p>Cookie value is: {cookie}</p>;
 };
diff --git a/packages/components/src/HookForm.tsx b/packages/components/src/HookForm.tsx
index 36eed3a..418a330 100644
--- a/packages/components/src/HookForm.tsx
+++ b/packages/components/src/HookForm.tsx
@@ -3,7 +3,7 @@ import { useForm } from "react-hook-form";
 
 export const HookForm: React.FC = () => {
   // The next line works with Next.js v11, but breaks with v12.
-  // useForm();
+  useForm();
 
   return <form>
     <label>Please fill the form</label>

```

Somehow the symlink created with preconstruct makes the trouble.

PS: using `experimental.externalDir = true` in `next.config.js` instead of `@preconstruct/next` seams to works. See
- https://github.com/vercel/next.js/blob/a803638da2483f0c468522d5da1037657af53e05/packages/next/build/webpack-config.ts#L951
  - https://github.com/vercel/next.js/discussions/31928
- https://github.com/preconstruct/preconstruct/blob/2cbbd3e8f4de1e7247610085b62763cdec448bcb/packages/next/index.js#L19

**Comment by anttimaki** [comment]
Appreciate your comments @namjul 

> With the following patch it works.
> ...

I'm not sure if the strike-trough comment below the code diff meant that this does not work, but at least for me it doesn't. Editing the `package.json` causes error when running `yarn install`:

> $ preconstruct dev
� error components main field is invalid, found `"src/index.ts"`, expected `"dist/components.cjs.js"`
� error components module field is invalid, found `"src/index.ts"`, expected `"dist/components.esm.js"`
� info Some of the errors above can be fixed automatically by running preconstruct fix
� info If want to learn more about the above error, check https://preconstruct.tools/errors
� info If the error is not there and you want to learn more about it, open an issue at https://github.com/preconstruct/preconstruct/issues/new
error Command failed with exit code 1.

Running `yarn preconstruct fix` just reverts the changes from `package.json`.

---

> PS: using experimental.externalDir = true in next.config.js instead of @preconstruct/next seams to works. See
...

Care to elaborate on this? Changing `next.config.js` to:

```js
/**
 * @type {import('next').NextConfig}
 */
const nextConfig = {
  experimental: {
    externalDir: true
  }
};

module.exports = nextConfig
```

Doesn't seem to change much to me. Without `apps/nextjs/babel.config.js` the compiler still seems to use SWC and things break down as they did before. Anyway, having a separate Babel config file is not an issue for me, I'm more interested in seeing if SWC could be used with Preconstruct.

**Comment by namjul** [comment]
> I'm not sure if the strike-trough comment below the code diff meant that this does not work, but at least for me it doesn't. Editing the package.json causes error when running yarn install

So what I meant with "working" was that if the `main`/`module` path goes directly to the entry file the nextjs app runs without error.
Since `yarn install` and therefore `preconstruct dev` will change that path back its not really a solution. Just a hint that it has something to do with the symlink created by preconstruct.

> Doesn't seem to change much to me. 

This was just a side-note that `@preconstruc/next` maybe can be replaced with `externDir: true`. Not really related to the original issue you brought up.

> Anyway, having a separate Babel config file is not an issue for me, I'm more interested in seeing if SWC could be used with Preconstruct.

I think it does if you don't use `preconstruct dev`, meaning always build the files first with `preconstruct build` or `preconstruct watch`. From my POV its about using SWC with `preconstruct dev` that is the issue.

**Comment by mitchellhamilton** [comment]
This has been fixed by 8b740fe81e55a3ca0395b36b55a949544147c2cc which was released in `@preconstruct/next@4.0.0`

**Comment by airtonix** [comment]
> This has been fixed by [8b740fe](https://github.com/preconstruct/preconstruct/commit/8b740fe81e55a3ca0395b36b55a949544147c2cc) which was released in `@preconstruct/next@4.0.0` (if it doesn't work immediately after upgrading, try deleting your `.next` directory)

Sorry, but what precisely has been fix? 

1. babel config still works as before pre nextjs@11, or
2. can use `preconstruct dev` without babel? (this is what I want for speed up improvements)
